packages/headless/src/controllers/insight/attached-citations/headless-attached-citations.ts

```diff
@@ -0,0 +1,188 @@
+ * @category AttachedCitations
+ */
+export function buildAttachedCitations(
+  engine: InsightEngine,
+  props: AttachedCitationsProps
+): AttachedCitations {
+  if (!loadAttachedResultsReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+  const {caseId} = props.options;
+
+  const isCitationAttached = (citation: GeneratedAnswerCitation): boolean => {
+    if (isNullOrUndefined(caseId)) {
+      return false;
+    }
+
+    if (
+      isNullOrUndefined(citation.permanentid) &&
+      isNullOrUndefined(citation.id)
+    ) {
+      return false;
+    }
+
+    return engine.state.attachedResults.results.some((attached) => {
+      const caseIdMatches = attached.caseId === caseId;
+      const permanentIdMatches =
+        !isNullOrUndefined(citation.permanentid) &&
+        attached.permanentId === citation.permanentid;
+      const uriHashMatches =
+        !isNullOrUndefined(citation.fields?.urihash) &&
+        attached.uriHash === citation.fields?.urihash;
+      const isCitation = attached.isCitation === true;
+
+      return (
+        caseIdMatches && isCitation && (permanentIdMatches || uriHashMatches)
+      );
+    });
+  };
+
+  const getAttachedResultsForRecord = (): AttachedResult[] => {
+    return engine.state.attachedResults.results.filter(
+      (attached) => attached.caseId === caseId && attached.isCitation === true
+    );
+  };
+
+  const validateStringOrUndefined = (value: unknown): string | undefined => {
+    return typeof value === 'string' ? value : undefined;
+  };
+
+  const buildAttachedResultFromCitation = (
+    citation: GeneratedAnswerCitation
+  ): AttachedResult => {
+    return {
+      caseId,
+      permanentId: citation.permanentid,
+      resultUrl: citation.clickUri ?? citation.uri,
+      title: citation.title,
+      uriHash: validateStringOrUndefined(citation.fields?.urihash),
+      source:
+        citation.source ?? validateStringOrUndefined(citation.fields?.source),
+      // Fields needed to support case articles
+      knowledgeArticleId: validateStringOrUndefined(
+        citation.fields?.knowledgeArticleId ?? citation.fields?.sfkbid
+      ),
+      articleLanguage: validateStringOrUndefined(citation.fields?.sflanguage),
+      articleVersionNumber: validateStringOrUndefined(
+        citation.fields?.sfversionnumber
+      ),
+      articlePublishStatus: validateStringOrUndefined(
+        citation.fields?.sfpublishstatus
+      ),
+      isCitation: true,
+    };
+  };
+
+  return {
+    ...controller,
+
+    get state() {
+      return {
+        results: getAttachedResultsForRecord(),
+      };
+    },
+
+    isAttached(citation: GeneratedAnswerCitation): boolean {
+      return isCitationAttached(citation);
+    },
+
+    attach(citation: GeneratedAnswerCitation): void {
+      const resultToAttach = buildAttachedResultFromCitation(citation);
+      dispatch(attachResult(resultToAttach));
```

Comment:
You can omit the intermediate variable if you want, only if you want.
```suggestion
      dispatch(attachResult(buildAttachedResultFromCitation(citation)));
```